lib/sha256.c

```diff
@@ -66,6 +66,33 @@ static void SHA256_Final(unsigned char digest[32], SHA256_CTX *ctx)
   sha256_digest(ctx, SHA256_DIGEST_SIZE, digest);
 }
 
+#elif defined(USE_GNUTLS)
+
+#include "curl_memory.h"
+
+/* The last #include file should be: */
+#include "memdebug.h"
+
+typedef struct gcry_md_hd_t SHA256_CTX;
+
+static int SHA256_Init(SHA256_CTX *ctx)
+{
+  gcry_md_open(ctx, GCRY_MD_SHA256, 0);
+}
+
+static void SHA256_Update(SHA256_CTX *ctx,
+                          const unsigned char *input,
+                          unsigned int inputLen)
+{
+  gcry_md_write(*ctx, input, inputLen);
+}
+
+static void SHA256_Final(unsigned char digest[32], SHA256_CTX *ctx)
+{
+  memcpy(digest, gcry_md_read(*ctx, 0), SHA256_DIGEST_LENGTH);
+  gcry_md_close(*ctx);
+}
+
```

Comment:
According to GnuTLS docs: *"The Cryptographic library layer, currently supports only libnettle. Older versions of GnuTLS used to support libgcrypt"* - so my question is: can you test this change? We don't in the CI. I'm thinking maybe this shouldn't be modified if we can't...

Reply:
> According to GnuTLS docs: _"The Cryptographic library layer, currently supports only libnettle. Older versions of GnuTLS used to support libgcrypt"_ - so my question is: can you test this change?

Good question... As you know I'm still a bit of *nix notice but I have performed the following on one of my Ubuntu 18.04 WSL installs:

- Modified configure.ac to force gcrypt rather than use nettle. Essentially I ended up with:

````
if test "$GNUTLS_ENABLED" = "1"; then
  USE_GNUTLS_NETTLE=
  LIBS="-lgcrypt $LIBS"
fi
````

- Reran ./buildconf
- Ran a modified version of my usual `./configure` to enable GNU TLS:

````
./configure --disable-shared --enable-debug --enable-curldebug --enable-maintainer-mode --enable-warnings --with-gnutls --without-ssl --enable-ldap --with-gssapi --with-libidn --with-rtmp --with-libssh2
````

- I then had some compilation errors as I had missing `#include <gcrypt.h>` :( which I fixed in a rebase.

- Using `typedef struct gcry_md_hd_t SHA256_CTX;` causes more compilation issues but changing it to `typedef gcry_md_hd_t SHA256_CTX;` worked. Note: md4.c has this issue so I'll create a PR for that. I note that `src/tool_metalink.c` uses `#define`s instead.

> We don't in the CI. I'm thinking maybe this shouldn't be modified if we can't...

We have this support for gcrypt in `lib/md4.c` and `src/tool_metalink.c`.